### DISCUSSION TITLE
mg_wakeup cannot be used with FreeRTOS

### DISCUSSION THREAD

**Discussion by tianlei-zhang** [cesanta/mongoose#2961] [Ask questions, get answers] [answered]
Hi,
I am trying to realize soft of multi-threading on FreeRTOS and found that mg_wakeup function does not exist if the architecture is set to FreeRTOS. It looks like the function is dependent on socket system support. Is this true?

**Comment by scaprile** [comment, marked as answer]
Please see our [documentation](https://mongoose.ws/documentation/),  and follow the guidelines in our [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples).
From https://mongoose.ws/documentation/tutorials/core/multi-threaded :
> To be able to use several threads, we need a mechanism that enables these threads to communicate with the main context in which the event manager is running. We can use a [lock-free queue](https://mongoose.ws/documentation/#queue) for that, though the event handler needs to poll for messages in that queue when it receives an MG_EV_POLL event. To have the event manager thread sleeping and wakeup when necessary, _Mongoose provides a **socket-based** helper function._

**Comment by tianlei-zhang** [comment]
Thanks Scaprile for the quick reply. This gives me a direction to explore.
